webrender/res/clip_shared.glsl

```diff
@@ -3,63 +3,55 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-flat varying vec4 vClipRect;
-flat varying vec4 vClipRadius;
-flat varying vec4 vClipMaskUvRect;
-flat varying vec4 vClipMaskLocalRect;
-
 #ifdef WR_VERTEX_SHADER
-void write_clip(ClipData clip) {
-    vClipRect = vec4(clip.rect.rect.xy, clip.rect.rect.xy + clip.rect.rect.zw);
-    vClipRadius = vec4(clip.top_left.outer_inner_radius.x,
-                       clip.top_right.outer_inner_radius.x,
-                       clip.bottom_right.outer_inner_radius.x,
-                       clip.bottom_left.outer_inner_radius.x);
-    //TODO: interpolate the final mask UV
-    vec2 texture_size = textureSize(sMask, 0);
-    vClipMaskUvRect = clip.mask_data.uv_rect / texture_size.xyxy;
-    vClipMaskLocalRect = clip.mask_data.local_rect; //TODO: transform
-}
-#endif
 
-#ifdef WR_FRAGMENT_SHADER
-float do_clip(vec2 pos) {
-    vec2 ref_tl = vClipRect.xy + vec2( vClipRadius.x,  vClipRadius.x);
-    vec2 ref_tr = vClipRect.zy + vec2(-vClipRadius.y,  vClipRadius.y);
-    vec2 ref_br = vClipRect.zw + vec2(-vClipRadius.z, -vClipRadius.z);
-    vec2 ref_bl = vClipRect.xw + vec2( vClipRadius.w, -vClipRadius.w);
+struct CacheClipInstance {
+    int render_task_index;
+    int layer_index;
+    int data_index;
+    int pad;
+};
+
+CacheClipInstance fetch_clip_item(int index) {
+    CacheClipInstance cci;
+
+    int offset = index * 1;
 
-    float d_tl = distance(pos, ref_tl);
-    float d_tr = distance(pos, ref_tr);
-    float d_br = distance(pos, ref_br);
-    float d_bl = distance(pos, ref_bl);
+    ivec4 data0 = int_data[offset + 0];
 
-    float pixels_per_fragment = length(fwidth(pos.xy));
-    float nudge = 0.5 * pixels_per_fragment;
-    vec4 distances = vec4(d_tl, d_tr, d_br, d_bl) - vClipRadius + nudge;
+    cci.render_task_index = data0.x;
+    cci.layer_index = data0.y;
+    cci.data_index = data0.z;
+    cci.pad = 0;
 
-    bvec4 is_out = bvec4(pos.x < ref_tl.x && pos.y < ref_tl.y,
-                         pos.x > ref_tr.x && pos.y < ref_tr.y,
-                         pos.x > ref_br.x && pos.y > ref_br.y,
-                         pos.x < ref_bl.x && pos.y > ref_bl.y);
+    return cci;
+}
+
+// The transformed vertex function that always covers the whole tile with the primitive
```

Comment:
Let's specify that the whole tile means the intersection of the clip instances here.